In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!pip install finance-datareader

In [ ]:
import pandas as pd
import numpy as np
import os
import FinanceDataReader as fdr

from sklearn.linear_model import LinearRegression
from tqdm import tqdm

## Get Stock List

In [ ]:
path = '/content/drive/MyDrive/Colab Notebooks/k-digital 3차/dacon_stock'
list_name = 'Stock_List.csv'
sample_name = 'sample_submission_week4.csv'

#csv에서 stock list 불러오기
stock_list = pd.read_csv(os.path.join(path,list_name))
stock_list['종목코드'] = stock_list['종목코드'].apply(lambda x : str(x).zfill(6))
stock_list

종목명    종목코드    상장시장
0        삼성전자  005930   KOSPI
1      SK하이닉스  000660   KOSPI
2       NAVER  035420   KOSPI
3         카카오  035720   KOSPI
4    삼성바이오로직스  207940   KOSPI
..        ...     ...     ...
371   더네이쳐홀딩스  298540  KOSDAQ
372       코엔텍  029960  KOSDAQ
373     원익홀딩스  030530  KOSDAQ
374       웹케시  053580  KOSDAQ
375    신흥에스이씨  243840  KOSDAQ

[376 rows x 3 columns]

## Get Data & Modeling

In [ ]:
start_date = '20210104'
end_date = '20210910'

#start_date 요일 숫자로 받아오기
start_weekday = pd.to_datetime(start_date).weekday()

#월요일을 첫째 요일이라는 기준으로 end_date의 주차를 10진수 str으로 변환
max_weeknum = pd.to_datetime(end_date).strftime('%V')

#start ~ end_date 영업일 저장
Business_days = pd.DataFrame(pd.date_range(start_date,end_date,freq='B'), columns = ['Date'])

print(f'WEEKDAY of "start_date" : {start_weekday}')
print(f'NUM of WEEKS to "end_date" : {max_weeknum}')
print(f'HOW MANY "Business_days" : {Business_days.shape}', )
display(Business_days.head())

WEEKDAY of "start_date" : 0
NUM of WEEKS to "end_date" : 36
HOW MANY "Business_days" : (180, 1)


Date
0 2021-01-04
1 2021-01-05
2 2021-01-06
3 2021-01-07
4 2021-01-08

## Baseline 모델의 구성 소개 ( Sample )

- X : (월 ~ 금) * 33주간
- y : (다음주 월 ~ 금) * 33주간
    - y_0 : 다음주 월요일
    - y_1 : 다음주 화요일
    - y_2 : 다음주 수요일
    - y_3 : 다음주 목요일
    - y_4 : 다음주 금요일


- 이번주 월~금요일의 패턴을 학습해 다음주 월요일 ~ 금요일을 각각 예측하는 모델을 생성
    
- 이 과정을 모든 종목(376개)에 적용

In [ ]:
sample_code = stock_list.loc[0,'종목코드']

#기간동안의 종가('Close') 데이터만 가져오기
sample = fdr.DataReader(sample_code, start = start_date, end = end_date)[['Close']].reset_index()

#sample에 없는 Business_days 를 Date에 추가 (Close 값은 없음)
sample = pd.merge(Business_days, sample, how = 'outer')

#날짜에 해당하는 요일값 'weekday' column에 추가
sample['weekday'] = sample.Date.apply(lambda x : x.weekday())

#주차를 'weeknum' column에 추가
sample['weeknum'] = sample.Date.apply(lambda x : x.strftime('%V'))

#인덱스상 앞의 데이터로 NULL 채움
sample.Close = sample.Close.ffill()

#pivot
sample = pd.pivot_table(data = sample, values = 'Close', columns = 'weekday', index = 'weeknum')
sample.head()

weekday        0        1        2        3        4
weeknum                                             
01       83000.0  83900.0  82200.0  82900.0  88800.0
02       91000.0  90600.0  89700.0  89700.0  88000.0
03       85000.0  87000.0  87200.0  88100.0  86800.0
04       89400.0  86700.0  85600.0  83700.0  82000.0
05       83000.0  84400.0  84600.0  82500.0  83500.0

In [ ]:
#모델 저장
model = LinearRegression()

In [ ]:
#최근 2주차만 남겨두고 x에 ndarray로 가져오기
x = sample.iloc[0:-2].to_numpy()

#34주차 데이터임을 확인
x.shape

(34, 5)

In [ ]:
#y에는 최근 1주차만 남겨두고 ndarray로 가져온다 
y = sample.iloc[1:-1].to_numpy()

#y_x 변수에 각각 요일 별 주차데이터를 저장
y_0 = y[:,0]
y_1 = y[:,1]
y_2 = y[:,2]
y_3 = y[:,3]
y_4 = y[:,4]

#리스트에 요일별 주차데이터를 저장
y_values = [y_0, y_1, y_2, y_3, y_4]

- 8월 마지막 주 예측

In [ ]:
#마지막 주 x_public에 저장
x_public = sample.iloc[-2].to_numpy()

array([74600., 76700., 76800., 76000., 76600.])

- 예측

In [ ]:
#요일별로 
predictions = []
for y_value in y_values :
    #요일별로 학습
    model.fit(x,y_value)
    #x_public에 차원추가 해서 예측
    prediction = model.predict(np.expand_dims(x_public,0))
    predictions.append(prediction[0])
predictions

[77302.1323626224,
 77260.59420262497,
 76717.45444247499,
 76882.61533987035,
 76437.26776807781]

- 실제 Public 값

In [ ]:
#실제 값을 통해 예측값에 어느 정도 차이가 있는지 확인
sample.iloc[-1].values

array([77300., 76100., 76300., 75300., 75300.])

# 전체 모델링

In [ ]:
#제출 csv 불러오기
sample_name = 'sample_submission_week4.csv'
sample_submission = pd.read_csv(os.path.join(path,sample_name))

In [ ]:
#sample에서 했던 것과 같은 방법으로 제출파일 속 기업들로 모두 학습 & 예측 시행
model = LinearRegression()
for code in tqdm(stock_list['종목코드'].values):
    data = fdr.DataReader(code, start = start_date, end = end_date)[['Close']].reset_index()
    data = pd.merge(Business_days, data, how = 'outer')
    data['weekday'] = data.Date.apply(lambda x : x.weekday())
    data['weeknum'] = data.Date.apply(lambda x : x.strftime('%V'))
    data.Close = data.Close.ffill()
    data = pd.pivot_table(data = data, values = 'Close', columns = 'weekday', index = 'weeknum')
    
    x = data.iloc[0:-2].to_numpy() # 2021년 1월 04일 ~ 2021년 8월 13일까지의 데이터로
    y = data.iloc[1:-1].to_numpy() # 2021년 1월 11일 ~ 2021년 8월 20일까지의 데이터를 학습한다.
    y_0 = y[:,0]
    y_1 = y[:,1]
    y_2 = y[:,2]
    y_3 = y[:,3]
    y_4 = y[:,4]

    y_values = [y_0, y_1, y_2, y_3, y_4]
    x_public = data.iloc[-2].to_numpy() # 2021년 8월 23일부터 8월 27일까지의 데이터를 예측할 것이다.
    
    predictions = []
    for y_value in y_values :
        model.fit(x,y_value)
        prediction = model.predict(np.expand_dims(x_public,0))
        predictions.append(prediction[0])
    sample_submission.loc[:,code] = predictions * 2
sample_submission.isna().sum().sum()

100%|██████████| 376/376 [07:04<00:00,  1.13s/it]


0

In [ ]:
sample_submission.columns

Index(['Day', '000060', '000080', '000100', '000120', '000150', '000240',
       '000250', '000270', '000660',
       ...
       '330860', '336260', '336370', '347860', '348150', '348210', '352820',
       '357780', '363280', '950130'],
      dtype='object', length=377)

In [ ]:
columns = list(sample_submission.columns[1:])

In [ ]:
columns = ['Day'] + [str(x).zfill(6) for x in columns]

In [ ]:
sample_submission.columns = columns

In [ ]:
#
sample_submission.to_csv('BASELINE_Linear_Week4.csv',index=False)

In [ ]:
#제출파일 미리 확인
sample_submission

Day        000060  ...        363280        950130
0  2021-09-06  29187.850221  ...  29824.046596  17783.411804
1  2021-09-07  29629.848350  ...  30030.835535  17505.425336
2  2021-09-08  31405.583931  ...  30060.148238  17568.946859
3  2021-09-09  31202.759683  ...  29896.565856  17553.054233
4  2021-09-10  30482.150750  ...  29829.829098  17461.664762
5  2021-09-27  29187.850221  ...  29824.046596  17783.411804
6  2021-09-28  29629.848350  ...  30030.835535  17505.425336
7  2021-09-29  31405.583931  ...  30060.148238  17568.946859
8  2021-09-30  31202.759683  ...  29896.565856  17553.054233
9  2021-10-01  30482.150750  ...  29829.829098  17461.664762

[10 rows x 377 columns]

!pip install dacon_submit_api-0.0.4-py3-none-any.whl

In [ ]:
from dacon_submit_api import dacon_submit_api 
파일경로 =  './BASELINE_Linear_Week3.csv'
개인_Token = ''                      # 개인 토큰 발급 (https://dacon.io/account 에서 발급 가능)
대회ID = '235800'                    # 주식종가예측 대회 ID 
팀이름 = ''                          # 현 대회에 참여중인 팀 명 (https://dacon.io/competitions/official/235800/team 에서 확인 가능)
submission_메모_내용 = 'sample_submission'

result = dacon_submit_api.post_submission_file(
파일경로, 
개인_Token, 
대회ID, 
팀이름, 
submission_메모_내용)